ath79: add support for AVM FRITZ!WLAN Repeater DVB-C

sindastra opened the pull request:
ath79: add support for AVM FRITZ!WLAN Repeater DVB-C

This commit adds support for the AVM FRITZ!WLAN Repeater DVB-C

SOC: Qualcomm Atheros QCA9556 720/600/200MHz MIPS 74Kc
RAM: 64 MiB (0x4000000)
FLASH: 16 MB NOR (WINBOND Uniform-Flash, 256 Bytes WriteBuffer)
WLAN: 2.4 GHZ b/g/n and 5 GHz n/ac
Ethernet: Gigabit
DVB-C: em28174 with MxL251 tuner
INPUT: 1x WPS Button (used for reset in OEM firmware by 15 seconds hold)
LED: Power, WLAN, TV, RSSI0-4
Serial: Yes

Tested and working:
 - Ethernet (correct MAC, gigabit, iperf3 about 200 Mbit/s)
 - 2.4 GHz Wi-Fi (correct MAC)
 - 5 GHz Wi-Fi (correct MAC)
 - WPS Button (tested using wifitoggle)
 - LEDs
 - Installation via EVA bootloader (FTP recovery)
 - OpenWrt sysupgrade (both CLI and LuCI)
 - Download of "urlader" (mtd0)

Not working:
 - Internal USB
 - DVB-C em28174+MxL251 (depends on internal USB)

Installation via EVA bootloader (FTP recovery):
Set NIC to 192.168.178.3/24 gateway 192.168.178.1 and power on the device,
connect to 192.168.178.1 through FTP and sign in with adam2/adam2:

ftp> quote USER adam2
ftp> quote PASS adam2
ftp> binary
ftp> debug
ftp> passive
ftp> quote MEDIA FLSH
ftp> put openwrt-sysupgrade.bin mtd1

Wait for "Transfer complete" together with the transfer details.
Wait two minutes to make sure flash is complete (just to be safe).
Then restart the device (power off and on) to boot into OpenWrt.
Revert your NIC settings to reach OpenWrt at 192.168.1.1

Signed-off-by: Natalie Kagelmacher <nataliek@pm.me>

adrianschmutzler commented:
For whatever reason, you have renamed your commit so it doesn't match the requirements anymore :-)

Despite, the mail address for Signed-off-by should be in brackets `<>`.

adrianschmutzler commented on target/linux/ath79/dts/qca9556_avm_fritzdvbc.dts:
```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include <dt-bindings/gpio/gpio.h>
```
This is already in the dtsi and can be dropped here.

adrianschmutzler commented on target/linux/ath79/dts/qca9556_avm_fritzdvbc.dts:
```diff
@@ -0,0 +1,108 @@
+#include "qca9556_avm_fritz-repeater.dtsi"
+
+/ {
+	compatible = "avm,fritzdvbc", "qca,qca9556";
```
I personally don't like these abbreviated names. I'd go for something really resembling the device name, e.g. avm,fritz-repeater-dvb-c ...

blocktrron replied:
Yes, however this is currently the case for all those boards. AVM also changed the name of their scorpion repeater 2 times so I'm fine with that as well as the machine name here.

adrianschmutzler replied:
I'm aware and I'm not too fond of that fritz-exception to naming in ath79. However, life's hard and I won't force anybody to change that now for a new device, as your argument is obviously valid.
Despite, it would be ugly to resemble that fritz!wlan naming anyway ...

adrianschmutzler commented on target/linux/ath79/dts/qca9556_avm_fritzdvbc.dts:
```diff
@@ -0,0 +1,108 @@
+		wlan {
+			label = "fritzdvbc:green:wlan";
+			gpios = <&spi_gpio 6 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy1tpt";
```
Based on the label, there is only one LED and you are "arbitrarily" selecting one WiFi for it here?
Or is this selection inherited from vendor firmware?

blocktrron replied:
Selecting the 2.4 radio as default is commonplace as we can not bind two radio triggers on one LED.

adrianschmutzler replied:
I typically prefer to have nothing set in this case, so the user can set it up the way he/she likes. Having LED activity by default would trick the uneducated user to assume vendor behavior (i.e. both WiFi are relevant) is implemented.
However, I'm aware that you could bring up a similar argument for your side, and after all it's not really important and a matter of taste anyway.

adrianschmutzler commented on target/linux/ath79/dts/qca9556_avm_fritzdvbc.dts:
```diff
@@ -0,0 +1,108 @@
+		line-name = "PCIE EP reset";
+	};
+
+	reset-pcie {
```
"reset-pcie-bus"?

sindastra replied:
What about it? The device has PCIe and uses it (IIRC for one of the Wi-Fi bands)

sindastra replied:
I should maybe also add that I got it from the OEM source code.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -25,6 +25,15 @@ avm,fritz1750e)
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "$boardname:green:rssi4" "wlan1" "80" "100"
 	;;
+avm,fritzdvbc)
+	ucidef_set_led_netdev "lan" "LAN" "$boardname:green:tv" "eth0"
```
If it's labelled "tv", I don't think it should be used for lan?

sindastra replied:
Since, for now, I can't get DVB-C working (requires internal USB working) the LED would otherwise have no use. Suggestions are welcome. Consider that users could easily remove the LAN LED setting in LuCI. But I guess you're right that it could be confusing. Would you prefer it unused by default?

adrianschmutzler replied:
> Would you prefer it unused by default?

yes. LEDs meant for other purposes shouldn't be hijacked _by default_, but the default should resemble vendor behavior if possible. If somebody wants to use the tv led for LAN later, he/she could configure that manually afterwards.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,7 @@ ath79_setup_interfaces()
 	alfa-network,ap121f|\
 	aruba,ap-105|\
 	avm,fritz1750e|\
+	avm,fritzdvbc|\
```
Please apply alphabetic sorting correctly (might change if you adjust the compatible)

sindastra replied:
AFAIK numbers come before letters in most sorting methods.

adrianschmutzler replied:
Yes, so avm,fritzdvbc should be after avm,fritz300e ...

sindastra replied:
Oh, my comment was based on the snippet shown on GitHub where this wasn't visible. Sorry!

sindastra commented:
Oh, I thought I had copied the commit message including the title. I'll look into your suggestions. I did notice a strange bug right now related to the LEDs that needs to be looked into. Some time after running, it's like the LEDs (pins) all swap. 🤔

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -182,6 +182,16 @@ define Device/avm_fritz1750e
+  SOC := qca9556
+  IMAGE_SIZE := 15232k
+  DEVICE_MODEL := FRITZ!WLAN Repeater DVB-C
+  DEVICE_PACKAGES += rssileds kmod-ath9k kmod-ath10k-ct-smallbuffers \
```
kmod-ath9k is in default packages

sindastra commented:
Thanks for the suggestions. I'll fix things up (possibly tonight) and let you know when it's ready for review. I really hope another force push won't mess with things further! 🤞

sindastra commented:
Just a little update to let you know I haven't forgotten, just didn't get to it as initially planned but I will.

sindastra committed:
ath79: add support for AVM FRITZ!WLAN Repeater DVB-C

This commit adds support for the AVM FRITZ!WLAN Repeater DVB-C

SOC: Qualcomm Atheros QCA9556 720/600/200MHz MIPS 74Kc
RAM: 64 MiB (0x4000000)
FLASH: 16 MB NOR (WINBOND Uniform-Flash, 256 Bytes WriteBuffer)
WLAN: 2.4 GHZ b/g/n and 5 GHz n/ac
Ethernet: Gigabit
DVB-C: em28174 with MxL251 tuner
INPUT: 1x WPS Button (used for reset in OEM firmware by 15 seconds hold)
LED: Power, WLAN, TV, RSSI0-4
Serial: Yes

Tested and working:
 - Ethernet (correct MAC, gigabit, iperf3 about 200 Mbit/s)
 - 2.4 GHz Wi-Fi (correct MAC)
 - 5 GHz Wi-Fi (correct MAC)
 - WPS Button (tested using wifitoggle)
 - LEDs
 - Installation via EVA bootloader (FTP recovery)
 - OpenWrt sysupgrade (both CLI and LuCI)
 - Download of "urlader" (mtd0)

Not working:
 - Internal USB
 - DVB-C em28174+MxL251 (depends on internal USB)

Installation via EVA bootloader (FTP recovery):
Set NIC to 192.168.178.3/24 gateway 192.168.178.1 and power on the device,
connect to 192.168.178.1 through FTP and sign in with adam2/adam2:

ftp> quote USER adam2
ftp> quote PASS adam2
ftp> binary
ftp> debug
ftp> passive
ftp> quote MEDIA FLSH
ftp> put openwrt-sysupgrade.bin mtd1

Wait for "Transfer complete" together with the transfer details.
Wait two minutes to make sure flash is complete (just to be safe).
Then restart the device (power off and on) to boot into OpenWrt.
Revert your NIC settings to reach OpenWrt at 192.168.1.1

Signed-off-by: Natalie Kagelmacher <nataliek@pm.me>

sindastra commented:
OK, successfully force pushed. Please review the current version.

There's one question I have: OEM firmware uses the "WLAN" LED as binary status light, either on or off. Should I replicate this behavior? Currently, it also shows Wi-Fi activity on that LED.

adrianschmutzler commented:
> There's one question I have: OEM firmware uses the "WLAN" LED as binary status light, either on or off. Should I replicate this behavior? Currently, it also shows Wi-Fi activity on that LED.

That's up to you. Typically, activity indication is chosen here, even if that's not the case for vendor firmware.

sindastra commented:
I had a thought... Since it's not possible to bind RSSILEDS to both 2.4GHz and 5GHz, and since 2.4GHz is chosen in such a case, it might make sense to set the WLAN LED to 5GHz, otherwise 5GHz has no status whatsoever while 2.4GHz has a status twice. :thinking:

sindastra commented:
So, are we going to discuss the LED behavior or will someone merge this? If it's fine by you, you could merge it and I can create a patch for the LED later if desired.

blocktrron commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/blocktrron.git

blocktrron commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -341,9 +342,7 @@ ath79_setup_macs()
-	avm,fritz450e)
-		label_mac=$(fritz_tffs -n macwlan -i $(find_mtd_part "tffs (1)"))
-		;;
+	avm,fritz450e|\
```
This change seems to be by accident.

blocktrron commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -352,6 +351,9 @@ ath79_setup_macs()
 		;;
+	avm,fritzdvbc)
+		label_mac=$(fritz_tffs -n macwlan -i $(find_mtd_part "tffs (1)"))
+		;;
```
As this can be merged with the 450E and the 1750E.

blocktrron reviewed (commented):
I've fixed this in my staging tree, so this is just FTR.
